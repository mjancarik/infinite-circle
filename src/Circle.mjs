import uid from 'easy-uid';
import arraysEqual from './arraysEqual.mjs';

export default class Circle {
  constructor({ listen, unlisten, filter, execute } = {}) {
    this.id = uid();

    this._listen = listen || (() => {});
    this._unlisten = unlisten || (() => {});
    this._filter = filter || this._defaultFilter;
    this._execute = execute || ((...args) => this._defaultExecute(...args));

    this._entries = new Map();
    this._lockCircle = false;

    this._observers = [];

    this._notifyWithContext = this.notify.bind(this);
  }

  subscribe(listener) {
    this._observers.push(listener);
  }

  notify(...args) {
    this._observers.map((listener) => {
      listener(...args);
    });
  }

  getEntries() {
    return this._entries;
  }

  register({ read, write, meta = { interval: 180 } }) {
    const id = uid();

    this._entries.set(id, {
      read,
      write,
      meta,
      time: Date.now() - meta.interval - 1,
      args: [],
    });

    if (this._entries.size === 1) {
      this._listen(this._notifyWithContext);
    }

    return id;
  }

  unregister(id) {
    this._entries.delete(id);

    if (this._entries.size === 0) {
      this._unlisten(this._notifyWithContext);
    }
  }

  filter(circleEntries, ...args) {
    return this._filter(circleEntries, ...args);
  }

  read(circleEntries, ...args) {
    const now = Date.now();

    return circleEntries.map((entry) => {
      entry.time = now;
      entry.args = args;
      const payload =
        typeof entry.read === 'function' ? entry.read(...args) : null;

      return Object.assign({ payload }, entry);
    });
  }

  write(circleEntries, ...args) {
    return circleEntries.map((entry) => {
      if (typeof entry.write === 'function') {
        return entry.write(entry, ...args);
      }
    });
  }

  *execute(...args) {
    if (this._lockCircle || this._entries.size === 0) {
      return [];
    }

    this._lockCircle = true;
    const modifiedEntries = yield* this._execute(...args);
    this._lockCircle = false;

    const originalEntries = Array.from(this.getEntries().values());
    const isCallAllEntriesWithArgs =
      originalEntries.filter((entry) => arraysEqual(entry.args, args))
        .length === originalEntries.length;

    if (!isCallAllEntriesWithArgs) {
      this.notify(...args);
    }

    return modifiedEntries;
  }

  _defaultFilter(circleEntries) {
    const now = Date.now();

    return Array.from(circleEntries.values()).filter((loopEntry) => {
      return loopEntry.time + loopEntry.meta.interval <= now;
    });
  }

  *_defaultExecute(...args) {
    let circleEntries = this.getEntries();

    circleEntries = this.filter(circleEntries, ...args);
    circleEntries = (yield circleEntries) || circleEntries;
    circleEntries = this.read(circleEntries, ...args);
    circleEntries = (yield circleEntries) || circleEntries;
    circleEntries = this.write(circleEntries, ...args);

    return circleEntries;
  }
}
