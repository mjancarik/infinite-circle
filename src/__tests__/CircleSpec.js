import Circle from '../Circle';
import { toMock } from 'to-mock';

describe('Circle', () => {
  let circle = null;

  const entry = {
    read() {
      return 1;
    },
    write() {
      return true;
    },
    meta: {
      interval: 60,
    },
    args: [],
    time: -60,
  };

  const args = { event: 'event' };
  const payload = 111;

  const NativeDate = Date;
  const MockedDate = toMock(Date);

  beforeEach(() => {
    jest.spyOn(MockedDate, 'now').mockReturnValue(1);
    jest.spyOn(MockedDate.prototype, 'getTime').mockReturnValue({
      toString() {
        return 'random';
      },
    });

    circle = new Circle();
    Date = MockedDate; // eslint-disable-line no-global-assign
  });

  afterEach(() => {
    Date = NativeDate; // eslint-disable-line no-global-assign
  });

  it('should call listen method for registering first entry', () => {
    jest.spyOn(circle, '_listen').mockImplementation();

    circle.register(entry);

    expect(circle._listen).toHaveBeenCalled();
    expect(circle.getEntries().size).toEqual(1);
  });

  it('should call unlisten method for unregistering last entry', () => {
    jest.spyOn(circle, '_unlisten').mockImplementation();

    const id = circle.register(entry);
    circle.unregister(id);

    expect(circle._unlisten).toHaveBeenCalled();
    expect(circle.getEntries().size).toEqual(0);
  });

  it('should call read method which return right structure for write method', () => {
    jest.spyOn(entry, 'read').mockReturnValue(payload);

    circle.register(entry);
    const entries = Array.from(circle.getEntries().values());

    expect(circle.read(entries, args)).toMatchSnapshot();
    expect(entry.read).toHaveBeenCalledWith(args);
  });

  it('should call write method with right structure', () => {
    jest.spyOn(entry, 'write').mockImplementation();

    circle.register(entry);
    const entries = Array.from(circle.getEntries().values()).map((entry) =>
      Object.assign({ payload }, entry)
    );

    circle.write(entries, args);

    expect(entry.write).toHaveBeenCalledWith(
      Object.assign({ payload }, entry),
      args
    );
  });

  describe('execute method', () => {
    function run(iterator) {
      let result = iterator.next();

      while (result.done === false) {
        result = iterator.next();
      }

      return result;
    }

    it('should return empty array for locked circle', () => {
      circle.register(entry);

      circle.execute().next();
      let iterator = circle.execute(args);
      let iterableObject = run(iterator);

      expect(iterableObject.value).toEqual([]);
      expect(iterableObject.done).toBeTruthy();
    });

    it('should return empty array for empty circle', () => {
      let iterator = circle.execute(args);

      let iterableObject = run(iterator);

      expect(iterableObject.value).toEqual([]);
      expect(iterableObject.done).toBeTruthy();
    });

    it('should return empty array if all entries are filtered', () => {
      MockedDate.now.mockReturnValue(60);
      circle.register(entry);
      let iterator = circle.execute(args);
      MockedDate.now.mockReturnValue(30);

      let iterableObject = run(iterator);

      expect(iterableObject.value).toEqual([]);
      expect(iterableObject.done).toBeTruthy();
    });

    it('should return empty array if all entries are filtered 2', () => {
      MockedDate.now.mockReturnValue(60);
      circle.register(entry);
      let iterator = circle.execute(args);
      MockedDate.now.mockReturnValue(61);

      let iterableObject = run(iterator);

      expect(iterableObject.value).toEqual([undefined]);
      expect(iterableObject.done).toBeTruthy();
    });

    it('should call notify method if all entries were not be called with defined arguments', () => {
      jest.spyOn(circle, 'notify').mockImplementation();
      MockedDate.now.mockReturnValue(60);
      circle.register(entry);
      let iterator = circle.execute(args);
      MockedDate.now.mockReturnValue(30);

      let iterableObject = run(iterator);

      expect(circle.notify).toHaveBeenCalledWith(args);

      expect(iterableObject.value).toEqual([]);
      expect(iterableObject.done).toBeTruthy();
    });
  });
});
