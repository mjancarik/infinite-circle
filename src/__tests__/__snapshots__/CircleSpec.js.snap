// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle should call read method which return right structure for write method 1`] = `
Array [
  Object {
    "args": Array [
      Object {
        "event": "event",
      },
    ],
    "meta": Object {
      "interval": 60,
    },
    "payload": 111,
    "read": [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "event": "event",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": 111,
        },
      ],
    },
    "time": 1,
    "write": [Function],
  },
]
`;
